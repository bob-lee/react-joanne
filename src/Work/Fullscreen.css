div#fullscreen {
  z-index: 213;
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.8);
  display: flex;
  align-items: center;
  justify-content: space-around;
  opacity: 0;
  visibility: hidden;
  /* transition-property: visibility, opacity, tranform;
  transition-duration: .4s, .4s, 0s;  */
  /* transition: visibility .4s linear, opacity .4s linear, transform 0s ease-in-out; */
}
div#fullscreen.show {
  opacity: 1;
  visibility: visible;
}

.fullscreen-enter {
  transform: translateX(100%);
}
.fullscreen-enter.left {
  transform: translateX(-100%);
}

.fullscreen-enter.fullscreen-enter-active {
  transform: translateX(0%);
  transition: transform .8s ease-in-out;
}

.fullscreen-exit {
  transform: translateX(0%);
}

.fullscreen-exit.fullscreen-exit-active {
  transform: translateX(-100%);
  transition: transform .8s ease-in-out;
}
.fullscreen-exit.fullscreen-exit-active.left {
  transform: translateX(100%);
  transition: transform .8s ease-in-out;
}
