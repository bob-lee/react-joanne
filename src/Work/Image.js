import React, { useState } from 'react'
import Loader from './Loader'

const Image = (props) => {
  const [loaded, setLoaded] = useState(false)

  let imgElement = null
  const handleLoad = (e) => {
    setLoaded(true)
    props.onImageLoaded({
      element: imgElement,
      index: props.index
    })
  }

  const item = props.item
  const classes = loaded ? 'image show' : 'image'
  
  const style = { margin: '0 auto' }
  const loading = (item.toLoad && !loaded) ? <Loader style={style} /> : null

  const thumbUrl = item.toLoad && item.thumbUrl
  const thumbImage = thumbUrl ? <img src={thumbUrl} className="image2" alt="thumb" /> : null

  return (
    <div>
      {loading}
      <div className={classes}>
        <a href={'#' + item.fileName} aria-hidden="true" className="anchor">
        <picture>
          <img src={item.toLoad ? item.url : undefined}
            onLoad={handleLoad}
            className="image1"
            ref={(el) => { imgElement = el; }}
            title="image"
            alt={item.fileName} />
        </picture>
        </a>

        <div className="expand">
          <span>{item.text}</span>
          {thumbImage}
        </div>
      </div>
    </div>
  )
}

export default Image
