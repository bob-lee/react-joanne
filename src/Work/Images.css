.icons .icon {
  z-index: 214;
  position: fixed;
  top: calc(50% - 30px);
  height: 60px;
  width: 60px;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 50%;
  background: white;
  visibility: hidden;
  opacity: 0;
  transition: visibility .4s linear, opacity .4s linear, transform .3s linear;
}
.icons.show .icon {
  visibility: visible;
  opacity: 0.3;
}

.icons .icon.exit {
  top: 1em;
  left: 5px;
}
.icons .icon.prev {
  left: 5px;
}
.icons .icon.next {
  right: 5px;
}
.icons .icon:hover {
  cursor: pointer;
  transform:scale(1.05);
}
