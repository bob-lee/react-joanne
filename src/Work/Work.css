.images {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  text-align: center;
  padding: 0.4em;
}

.image {
  display: block;
  width: 96vw;
  height: 100%; /* IE11 */
  padding-bottom: 1em;

  visibility: hidden;
  opacity: 0;
  transition: visibility 1.0s linear, opacity 1.0s linear;
}

.image.show {
  visibility: visible;
  opacity: 1;
}

.image-pair {
  position: relative;
}

.image1 {
  width: 100%;
}

@media only screen 
and (min-device-width: 768px) 
and (max-device-width: 1024px) 
and (orientation: portrait) 
and (-webkit-min-device-pixel-ratio: 2) {
  .image {
    width: 70vw;
  }
}
@media only screen 
and (min-device-width: 768px) 
and (max-device-width: 1024px) 
and (orientation: landscape) 
and (-webkit-min-device-pixel-ratio: 2) {
  .image {
    width: 52vw;
  }
}
@media only screen 
and (min-width: 640px) {
  .image {
    width: 40vw;
  }
}

.expand {
  display: flex;
  flex-flow: row;

  overflow: hidden;
}
.expand span {
  flex: 2;
}
.expand .image2 {
  height: 80px;
  width: 80px;
  border-radius: 50%;
}