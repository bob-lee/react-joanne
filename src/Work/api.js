import fetch from 'isomorphic-fetch'

const API = 'https://us-central1-joanne-lee.cloudfunctions.net/getUrls';

export default function getUrls(path) {
  return fetch(`${API}/${path}`)
    .then(response => {
      console.log(response)
      if (response.ok) {
        return response.json()
      } else {
        console.error('getUrls failed: ' + response.statusText)
      }
    })
    .catch(error => {
      throw new Error('getUrls error: ' +  error)
    })
}

