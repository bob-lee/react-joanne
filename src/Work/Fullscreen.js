import React, { useEffect, useState } from 'react'
if (typeof window !== 'undefined') {
  require('./Fullscreen.css')
}

const Fullscreen = ({img, toShow, onExit, onPrev, onNext}) => {
  const [fileName, setFileName] = useState('')

  useEffect(() => {
    const drawImage = (img) => {
      const canvas = document.getElementById('canvas');
      const ctx = canvas.getContext('2d');
      const scaled = getFullscreenSize(img.naturalWidth, img.naturalHeight)
  
      canvas.width = scaled.width
      canvas.height = scaled.height
      ctx.drawImage(img, 0, 0, scaled.width, scaled.height)
      // ctx.drawImage(img, 0, 0, sWidth, sHeight, 0, 0, scaled.width, scaled.height)
    }
  
    const getFullscreenSize = (sWidth, sHeight) => {
      const cWidth = document.documentElement.clientWidth
      const cHeight = document.documentElement.clientHeight
      console.log(`source(${sWidth} x ${sHeight})`)
      console.log(`client(${cWidth} x ${cHeight})`)
  
      let scaledWidth = scale(sWidth, sHeight, cWidth, cHeight, true)
      let scaledHeight = scale(sWidth, sHeight, cWidth, cHeight, false)
  
      let preferedScaled = scaledHeight
      let preferredWidth = false
      if (Math.abs(scaledWidth.overflow) < Math.abs(scaledHeight.overflow)) {
        preferedScaled = scaledWidth
        preferredWidth = true
      }
  
      console.log(`scaled once`, preferredWidth)
      let result = null
  
      if (cWidth >= preferedScaled.width && cHeight >= preferedScaled.height) {
        // got result
      } else { // scale once more
        if (cWidth < preferedScaled.width) {
          preferredWidth = false
          result = scale(preferedScaled.width, preferedScaled.height, cWidth, cHeight, preferredWidth)
        } else if (cHeight < preferedScaled.height) {
          preferredWidth = true
          result = scale(preferedScaled.width, preferedScaled.height, cWidth, cHeight, preferredWidth)
        }
      }
  
      if (result) {
        console.log(`scaled twice`, preferredWidth)
      } else {
        result = preferedScaled
      }
      logScale(result, preferedScaled)
  
      return result ? {
        width: result.width,
        height: result.height
      } : null
    }
  
    const scale = (wS, hS, wC, hC, scaleWidth) => {
      let overflow = 0
      let width = 0
      let height = 0
      if (scaleWidth) {
        const scaled = Math.floor(wS * hC / hS)
        overflow = (scaled - wC) * hC
        width = scaled
        height = hC
      } else {
        const scaled = Math.floor(hS * wC / wS)
        overflow = (scaled - hC) * wC
        width = wC
        height = scaled
      }
  
      const result = { width, height, overflow }
      logScale(result, scaleWidth)
      return result
    }
  
    const logScale = (scaled, scaleWidth) => console.log(`scaled${scaleWidth ? 'Width' : 'Height'}(${scaled.width} x ${scaled.height})`, scaled.overflow)
  
    if (img) {
      if (!fileName || fileName !== img.alt) {
        drawImage(img)
        setFileName(img.alt)
      }
    }

    console.log('Fulscreen', img && img.alt)
  }, [img])

  const classes = toShow ? 'show' : ''

  return(
    <div id="fullscreen" className={classes}>
      <i className="exit" onClick={onExit}>
        <svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="times" className="svg-inline--fa fa-times fa-w-11" role="img" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 352 512"><path fill="currentColor" d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"></path></svg>
      </i>
      <i className="prev" onClick={onPrev}>
        <svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="chevron-left" className="svg-inline--fa fa-chevron-left fa-w-10" role="img" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 320 512"><path fill="currentColor" d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"></path></svg>
      </i>
      <i className="next" onClick={onNext}>
        <svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="chevron-right" className="svg-inline--fa fa-chevron-right fa-w-10" role="img" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 320 512"><path fill="currentColor" d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"></path></svg>
      </i>
      <canvas id="canvas"></canvas>
    </div>
  )
}

export default Fullscreen
