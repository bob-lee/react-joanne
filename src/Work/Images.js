import React from 'react'
import Image from './Image'

const isWindow = typeof window !== 'undefined'
if (isWindow) {
  var scroll = require('react-scroll').animateScroll
}

class Images extends React.Component {
  state = { showIcon: false }
  lastY = 0
  ticking = false

  scrollToTop = () => scroll.scrollToTop()
  currentPositionY = () => window.pageYOffset

  updateLastY() {
    const newY = this.currentPositionY()
    if (newY !== this.lastY) {
      this.lastY = newY
      this.setState({
        showIcon: isWindow && this.lastY > 500
      })
      //console.log('updateLastY', Math.ceil(newY))
    }
  }

  handleScroll = (e) => {
    if(!this.ticking) {
      window.requestAnimationFrame(() => {
        this.updateLastY()
        this.ticking = false
      })
      this.ticking = true
    }
  }

  componentDidMount() {
    isWindow && window.addEventListener('scroll', this.handleScroll)
  }

  componentWillUnmount() {
    isWindow && window.removeEventListener('scroll', this.handleScroll)
  }

  render() {
    const { list, ...propsButList } = this.props
    const classes = `back-to-top ${this.state.showIcon ? 'show' : ''}`
    const info = Math.ceil(this.lastY)

    return (
      <div>
        <div className="images">
          {list.map((item, index) =>
            <Image key={item.fileName}
              item={item}
              index={index}
              {...propsButList} />
          )}
        </div>
        <div className={classes} onClick={this.scrollToTop}>
          <i className="fa fa-step-forward fa-lg" aria-hidden="true"></i>
        </div>
        <div className="back-to-top test">
          {info}
        </div>
      </div>
    )
  }
}

export default Images
