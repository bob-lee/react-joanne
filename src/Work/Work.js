import React from 'react';
import withObserver from './withObserver'
import Images from './Images'
import './Work.css'

const ImagesWithObserver = withObserver(Images)

export default ImagesWithObserver