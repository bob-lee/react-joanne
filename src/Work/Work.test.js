import React from 'react'
import {
  render,
  cleanup,
  waitForElement,
  getAllByAltText,
} from '@testing-library/react'
import 'jest-dom/extend-expect'
import Work from './Work'

global.IntersectionObserver = jest.fn()
jest.mock('./api')

const getProps = path => ( 
  { 
    match: { params: { name: path } },
    location: { hash: '' } 
  } 
)

afterEach(cleanup)

it('makes an initial API call and renders, first two images to have src provided', async () => {
  isTouchDevice(false)
  const props = getProps('painting')
  const {getAllByTitle, container} = render(
    <Work {...props} />,
  )

  const imageNodes = await waitForElement(() => getAllByTitle('image'))
  const thumbNodes = getAllByAltText(container, 'thumb')
  console.log(`found ${imageNodes.length} image(s)`)
  console.log(`found ${thumbNodes.length} thumb(s)`)

  expect(container.firstChild).toMatchSnapshot()
  expect(imageNodes.length).toBe(3)
  expect(imageNodes[0].src).toBeTruthy()
  expect(imageNodes[1].src).toBeTruthy()
  expect(imageNodes[2].src).toBeFalsy()
  expect(thumbNodes.length).toBe(2)
})
