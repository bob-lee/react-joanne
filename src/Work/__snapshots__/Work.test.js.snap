// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Work full-renders correctly 1`] = `
<Observer
  match={
    Object {
      "params": Object {
        "name": "painting",
      },
    }
  }
>
  <Images
    list={Array []}
    match={
      Object {
        "params": Object {
          "name": "painting",
        },
      }
    }
    onImageLoaded={[Function]}
  >
    <div>
      <div
        className="images"
      />
      <div
        className="back-to-top "
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="fa fa-step-forward fa-lg"
        />
      </div>
      <div
        className="back-to-top test"
      >
        0
      </div>
    </div>
  </Images>
</Observer>
`;

exports[`Work renders correctly 1`] = `
<div>
  <div
    className="images"
  />
  <div
    className="back-to-top "
    onClick={[Function]}
  >
    <i
      aria-hidden="true"
      className="fa fa-step-forward fa-lg"
    />
  </div>
  <div
    className="back-to-top test"
  >
    0
  </div>
</div>
`;

exports[`Work shallow-renders correctly 1`] = `
<Images
  list={Array []}
  match={
    Object {
      "params": Object {
        "name": "painting",
      },
    }
  }
  onImageLoaded={[Function]}
/>
`;
